include/swift/SIL/SILOpenedArchetypesTracker.h

```diff
@@ -0,0 +1,163 @@
+  void unregisterOpenedArchetypes(const SILInstruction *I);
+
+  // Handling of instruction removal notifications.
+  bool needsNotifications() { return true; }
+
+  // Handle notifications about removals of instructions.
+  void handleDeleteNotification(swift::ValueBase *Value);
+
+  virtual ~SILOpenedArchetypesTracker() {
+    // Unregister the handler.
+    F.getModule().removeDeleteNotificationHandler(this);
+  }
+
+private:
+  /// The function whose opened archetypes are being tracked.
+  /// Used only for verification purposes.
+  const SILFunction &F;
+  /// Mapping from opened archetypes to their definitions.
+  OpenedArchetypeDefsMap &OpenedArchetypeDefs;
+  /// Local map to be used if no other map was provided in the
+  /// constructor.
+  OpenedArchetypeDefsMap LocalOpenedArchetypeDefs;
+};
+
+// A state object containing information about opened archetypes.
+// This information can be used by constructors of SILInstructions,
+// their create methods, etc.
+// The object can be configured to use different sources for providing
+// archetypes, but none of those archetype sets can be modified through
+// this object, which makes it essentially immutable.
+class SILOpenedArchetypesState {
+  // A set of opened archetypes operands for quick lookup.
+  // It usually provides opened archetypes operands of the
+  // instruction being currenlty processed.
+  ArrayRef<Operand> OpenedArchetypeOperands;
+  // A non-modifiable mapping provided by the tracker.
+  const SILOpenedArchetypesTracker *OpenedArchetypesTracker;
+public:
+  SILOpenedArchetypesState(const SILOpenedArchetypesTracker *Tracker = nullptr)
+      : OpenedArchetypesTracker(Tracker) {}
+
+  SILOpenedArchetypesState(const SILOpenedArchetypesTracker &Tracker)
+      : OpenedArchetypesTracker(&Tracker) { }
+
+  void setOpenedArchetypesTracker(const SILOpenedArchetypesTracker *Tracker) {
+    OpenedArchetypesTracker = Tracker;
+  }
+
+  void addOpenedArchetypeOperands(ArrayRef<Operand> Operands) {
+    OpenedArchetypeOperands = Operands;
+  }
+
+  const SILOpenedArchetypesTracker *getOpenedArchetypesTracker() const {
+    return OpenedArchetypesTracker;
+  }
+
+  // Lookup the instruction defining an opened archetype by first 
+  // performing a quick lookup in the opened archetypes operands
+  // and then in the opened archetypes tracker.
+  SILValue getOpenedArchetypeDef(Type Ty) const;
+};
+
+/// Find an opened archetype defined by an instruction.
+/// \returns The found archetype or empty type otherwise.
+CanType getOpenedArchetypeOf(const SILInstruction *I);
+
+/// Find an opened archetype represented by this type.
+/// Typically, it would be called from a type visitor.
+/// It checks only the type itself, but does not try to
+/// recursively check any children of this type, because
+/// this is the task of the type visitor invoking it.
+/// \returns The found archetype or empty type otherwise.
+CanType getOpenedArchetypeOf(CanType Ty);
```

Comment:
In general, a type can involve multiple opened archetypes.  If this assumes that it is given e.g. the result type of an open instruction, it should probably assert that it's given either an opened archetype or a metatype tower thereof, and I wonder if it actually needs to be "public" this way.